specs/vep-2448-vdk-run-logs-simplified-and-readable/README.md

```diff
@@ -366,10 +366,290 @@ or slowed down due to progress bar updates. This is really relevant for
 ingestion.
 
 Also as we need progress indicators in multiple places likely we need a common
-encapsulation for that - in a form of a python module or plugin. The progress
+encapsulation (or abstraction) for that - in a form of a python module or plugin. The progress
 indicator should provide a notification callback mechanism or similar so it's
 integratable with Notebook
 
+For more detailed understanding of the user workflows [this research documented is recommended](https://github.com/vmware/versatile-data-kit/wiki/Research:-Progress-Indicators)
+
+#### Technical Requirements
+
+* The progress abstraction should support multi step operations. Different steps are started and ended multiple times one after another in a job.
+* The abstraction should support hierarchical operations. A Step can start other long running operations (e.g a DAG sub-jobs)
+* The abstraction should support operations running concurrently (that is multiple tasks are started and keep progressing concurrently - in the same thread).
+  For example, this requirement comes primarily because of DAGs. Multiple sub-jobs started in a DAG can run concurrently (no multi-threading needed here though)
+* The abstraction should support operations running in background threads. When you send data for ingestion, that data is processed in background while the job continue with other operations. The progress track can be updated from multiple threads as they ingest data
+* The abstraction should allow for the total to change. E.g when we are ingesting data the total number of objects being ingested is changed as user adds new and new ones.
```

Comment:
This is not quite clear to me. Either you know the total when you start ingesting, or you don't (in case of streams).

Reply:
The way usrs ingeset data is by using job_input.send_object_for_ingestion and job_input.send_tabular_data_for_ingestion 

And they can use it in any step in any place in any location. Those calls are asynchronous. That is send_xxx adds the payload to queue and returns immediately (there's config options to make them synchronous but nobody knows about it :) and unlikely to need it as it will slow the job ) 

That's why we have a requirement that the progress indicators must handle task running in background and also this one. Since the user can call send_tabular_data_for_ingestion at any time and that would add one more payload that need to be tracked.